udp/client/conn.go

```diff
@@ -744,8 +744,9 @@ func (cc *Conn) handlePong(w *responsewriter.ResponseWriter[*Conn], r *pool.Mess
 func (cc *Conn) handleSpecialMessages(r *pool.Message) bool {
 	// ping request
 	if r.Code() == codes.Empty && r.Type() == message.Confirmable && len(r.Token()) == 0 && len(r.Options()) == 0 && r.Body() == nil {
+		// handle ping request by sending pong, but still propogate the message
+		// to any handlers incase the application wants to know when pings happened
 		cc.ProcessReceivedMessageWithHandler(r, cc.handlePong)
```

Comment:
You shouldn't propagate the ping to the handler via `cc.handleSpecialMessages.C()` because the response has already been processed. Instead, you can extend the options related to the callback/channel that will be triggered after the request has been [unmarshalled line 782](https://github.com/plgd-dev/go-coap/blob/a48812ee4db5dbec9c8fb43496e8fc83dee79f7b/udp/client/conn.go#L782) in **the goroutine**.